Update transition script to allow docker/podman usage (#4427)
Azure/azure-sdk-tools: tools/test-proxy/scripts/transition-scripts/README.md

The script is [generate-assets-json.ps1](https://github.com/Azure/azure-sdk-tools/blob/main/tools/test-proxy/transition-scripts/generate-assets-json.ps1)

### Download the transition script locally

```powershell
Invoke-WebRequest -OutFile "generate-assets-json.ps1" https://raw.githubusercontent.com/Azure/azure-sdk-tools/main/tools/test-proxy/scripts/transition-scripts/generate-assets-json.ps1
```

```bash

> wget https://raw.githubusercontent.com/Azure/azure-sdk-tools/main/tools/test-proxy/scripts/transition-scripts/generate-assets-json.ps1

Reviewer: I would still add a specific outfile flag here, otherwise wget appends a number to the filename on re-download, which can be annoying if you are re-running the script by name and not realizing you aren't running the updated version.

```suggestion
wget https://raw.githubusercontent.com/Azure/azure-sdk-tools/main/tools/test-proxy/scripts/transition-scripts/generate-assets-json.ps1 -o generate-assets-json.ps1
```

```

## Setup

> Before running the script, understand that **only services that have migrated to use the `test-proxy` as their record/playback solution can store recordings into the external assets repository.** The test-proxy itself contains the code for `restoring`/`push`ing recordings, so if it is NOT being used for record/playback, that work must be completed before recordings can be moved.

Reviewer: I think the code blocks are confusing here, because `restoring` is not a test proxy verb. Maybe just remove them, or an alternative?

```suggestion
Before running the script, understand that **only services that have migrated to use the `test-proxy` as their record/playback solution can store recordings into the external assets repository.** The test-proxy itself contains the code for restoring/pushing recordings (`test-proxy [restore|push]`). If it is NOT being used for record/playback, that work must be completed before recordings can be moved.
```

> Running the script requires the base requirements.

Author: ```suggestion
Running the script requires these base requirements.
```

- [x] The targeted library is already migrated to use the test-proxy.
- [x] Git version `>2.25.0` needs to be on the machine and in the path. Git is used by the script and test-proxy.
- [x] [Powershell Core](https://learn.microsoft.com/en-us/powershell/scripting/install/installing-powershell?view=powershell-7.2) at least version 7.
- [x] Ensure global git config settings for `user.name` and `user.email` are updated. [Reference](https://git-scm.com/book/en/v2/Getting-Started-First-Time-Git-Setup)
  - Override with environment variables `GIT_COMMIT_EMAIL` and `GIT_COMMIT_OWNER`. If either of these are set, they will override the default values pulled from `git config --global`.

Once the above requirements are met, developers are welcome to choose one of the following paths.

### `test-proxy` dotnet tool installed and called directly

> Provide `TestProxyExe` argument of `test-proxy` or leave it **blank**. This is the default use-case of this transition script.

Reviewer: `-TestProxyExe` does not exist as a parameter in the script, it's just a global variable. Did you want to make it a parameter?

- [x] Test-proxy needs to be on the machine and in the path. Instructions for that are [here](https://github.com/Azure/azure-sdk-tools/blob/main/tools/test-proxy/Azure.Sdk.Tools.TestProxy/README.md#installation).

The newly installed test-proxy tool will be used during the recording migration portion of this script.

### `docker` or `podman` invocation

> To utilize this methodology, the user must set input argument `TestProxyExe` to `docker` or `podman`.

Reviewer: See above comment re the parameter.

Other requirements:

- [x] Install [docker](https://docs.docker.com/engine/install/) or [podman](https://podman.io/getting-started/installation.html)

> - [x] Set the environment variable `GIT_TOKEN` a valid token representing YOUR user

Reviewer: This is only a requirement now if you are using https auth. Don't know if you want to specify the ssh exception here as well, or just let people discover it on their own if they aren't using https.

Reviewer: Hm, I guess it's kind of more annoying because you'd have to inject it into the test proxy command somehow...

Reviewer: Nvm, see below comment, I think we can handle this (happy to do an update myself).

## Running the script

> [generate-assets-json.ps1](https://github.com/Azure/azure-sdk-tools/blob/main/tools/test-proxy/transition-scripts/generate-assets-json.ps1) is a standalone powershell script with no supporting script requirements. The easiest way to run the script would be to use a one-liner [defined above](#download-the-transition-script-locally) to grab the file directly. **Please ensure you have the newest version of this script before continuing!**

Reviewer: Lol, maybe the script should just download itself :P

```powershell
# if downloading the file singly, cd to the directory containing generate-assets-json.ps1
cd "<target-language-repo>/sdk/<service>"
<path-to-transition-script>/generate-assets-json.ps1
```

> The script needs to be executed inside an `sdk/<ServiceDirectory>` or deeper and from within an up to date language repository. A good rule here would be look at where the ci.yml is for an service directory. In the case where each library for a given service directory has their own pipelines, at the `sdk/<ServiceDirectory><Library>` level, it is recommended that the assets.json is created there. If the `ci.yml` exists deeper then the `sdk/<ServiceDirectory>/<Library>` level, then it is recommended to run the script from that directory.

Reviewer 2: ```suggestion
The script needs to be executed inside an `sdk/<ServiceDirectory>` or deeper and from within an up to date language repository. A good rule here would be look at where the ci.yml is for an service directory. In the case where each library for a given service directory has their own pipelines, at the `sdk/<ServiceDirectory>/<Library>` level, it is recommended that the assets.json is created there. If the `ci.yml` exists deeper than the `sdk/<ServiceDirectory>/<Library>` level, then it is recommended to run the script from that directory.
```

```powershell
# calling transition script against tool, given local clones of azure-sdk-for-java and azure-sdk-tools
cd c:/src/azure-sdk-for-java/sdk/attestation

> <path-to-transition-script>/generate-assets-json.ps1 -TestProxyExe "test-proxy" -InitialPush

Reviewer: I think you have good docs on how to override the exe above, so I would remove it here.

```
<path-to-transition-script>/generate-assets-json.ps1 -InitialPush
```

```

```powershell

> # calling transition script against tool, given local clones of azure-sdk-for-java and azure-sdk-tools

Reviewer 2: ```suggestion
# calling transition script against docker, given local clones of azure-sdk-for-java and azure-sdk-tools
```
